contracts/token/OtcEscrow.sol

```diff
@@ -0,0 +1,139 @@
+
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    /* ========== Events =========== */
+
+    event VestingDeployed(address vesting);
+    
+    /* ====== Modifiers ======== */
+
+    /**
+     * Throws if the sender is not Index Gov
+     */
+    modifier onlyIndexGov() {
+        require(msg.sender == indexGov, "unauthorized");
+        _;
+    }
+
+    modifier onlyApprovedParties() {
+        require(msg.sender == indexGov || msg.sender == beneficiary, "unauthorized");
+        _;
+    }
+
+    /* ======== State Variables ======= */
+
+    address public usdc;
+    address public index;
+
+    address public indexGov;
+    address public beneficiary;
+
+    uint256 public vestingStart;
+    uint256 public vestingEnd;
+    uint256 public vestingCliff;
+
+    uint256 public usdcAmount;
+    uint256 public indexAmount;
+
+
+
+    /* ====== Constructor ======== */
+
+    /**
+     * Sets the state variables that encode the terms of the OTC sale
+     *
+     * @param _beneficiary  Address that will purchase INDEX
+     * @param _indexGov     Address that will receive USDC
+     * @param _vestingStart Timestamp of vesting start
+     * @param _vestingCliff Timestamp of vesting cliff
+     * @param _vestingEnd   Timestamp of vesting end
+     * @param _usdcAmount   Amount of USDC swapped for the sale
+     * @param _indexAmount  Amount of INDEX swapped for the sale
+     * @param _usdcAddress  Address of the USDC token
+     * @param _indexAddress Address of the Index token
+     */
+    constructor(
+        address _beneficiary,
+        address _indexGov,
+        uint256 _vestingStart,
+        uint256 _vestingCliff,
+        uint256 _vestingEnd,
+        uint256 _usdcAmount,
+        uint256 _indexAmount,
+        address _usdcAddress,
+        address _indexAddress
+    ) public {
+        beneficiary = _beneficiary;
+        indexGov =  _indexGov;
+
+        vestingStart = _vestingStart;
+        vestingCliff = _vestingCliff;
+        vestingEnd = _vestingEnd;
+
+        usdcAmount = _usdcAmount;
+        indexAmount = _indexAmount;
+
+        usdc = _usdcAddress;
+        index = _indexAddress;
+    }
+    
+    /* ======= External Functions ======= */
+
+    /**
+     * Return INDEX to Index Governance to revoke the deal
+     */
+    function revoke() external onlyIndexGov {
+        uint256 indexBalance = IERC20(index).balanceOf(address(this));
+        IERC20(index).safeTransfer(indexGov, indexBalance);
+    }
+
+    /**
+     * Recovers USDC accidentally sent to the contract
+     */
+    function recoverUsdc() external {
+        uint256 usdcBalance = IERC20(usdc).balanceOf(address(this));
+        IERC20(usdc).safeTransfer(beneficiary, usdcBalance);
+    }
+
+    /**
+     * Executes the OTC deal. Sends the USDC from the beneficiary to Index Governance, and
+     * locks the INDEX in the vesting contract. Can only be called by approved parties.
+     */
+    function swap() external onlyApprovedParties {
```

Comment:
Can we order this function first? It's the most important one

Reply:
Sorry to dig this up again, but is there any vulnerability in allowing anyone to call swap?
Right now this would require either the Index Coop MultiSig (Alex / Felix) or the investor to call swap.
I'd like to reduce the number of transactions the Index Coop Multisig must execute, and I don't think asking investors to pay the swap / contract creation cost will add friction.

Ideally I would use another wallet to call swap and get reimbursed by the Treasury Committee.

Reply:
So the concern is that if the investor has an infinite approval for usdc, someone can call swap multiple times and draw extra usdc from their wallet. I think this is not a huge concern though since the swap function will just fail since their is not enough INDEX in the contract to complete the transaction. What do you think @bweick?

Reply:
Only thing I could think of here is that say (god forbid) INDEX really tanks in price someone could just deposit their INDEX here and pull USDC to the treasury. It's more of a griefing style attack since the INDEX depositor wouldn't directly benefit (unless it was the treasury itself) but it's still possible if the beneficiary does an unlimited approve.

Maybe we can remove it and when we produce instructions for the beneficiaries make sure to tell them not to do an infinite approve. But still don't like the idea of an approval hanging around out there that anyone could take advantage of.

Reply:
Other option here is to add an allow list of addresses that can call this function. That way it doesn't have to be a multi-sig and it's still a somewhat trusted address